src/protocol/mod.rs

```diff
@@ -119,20 +100,65 @@ impl Writer {
             Err(e) => Err(e),
         }
     }
+
+    #[inline]
+    fn is_empty(&self) -> bool {
```

Comment:
Instead of having `is_empty`, can we just have `WriteId` and `WriteSize` just contain an `Option<VecWriter>` and set it to `None` if the payload is empty?